lib/util/util.c

```diff
@@ -229,180 +118,62 @@ int make_dir_subdir(const char *parent, const char *dir)
 	return 0;
 }
 
-int attach_xdp_program(const struct bpf_object *obj, const char *prog_name,
-		       const struct iface *iface, bool force, enum xdp_attach_mode mode,
-		       const char *pin_root_path)
+int attach_xdp_program(struct xdp_program *prog, const struct iface *iface,
+		       enum xdp_attach_mode mode, const char *pin_root_path)
 {
-	char pin_path[PATH_MAX], old_prog_name[100];
-	struct bpf_program *prog = NULL;
-	int ifindex = iface->ifindex;
-	int err = 0, xdp_flags = 0;
-	struct bpf_prog_info info;
-	struct stat sb = {};
-	bool has_old;
-	int prog_fd;
-
-	has_old = program_is_loaded(iface->ifindex, NULL, NULL, &info);
-
-	if (!force) {
-		if (has_old) {
-			pr_warn("Program already loaded on %s; use --force to replace\n",
-				iface->ifname);
-			return -EEXIST;
-		}
-
-		xdp_flags |= XDP_FLAGS_UPDATE_IF_NOEXIST;
-	} else if (has_old) {
-		pr_debug("Replacing old program '%s' on iface '%s'\n",
-			 info.name, iface->ifname);
-	}
-
-	switch (mode) {
-	case XDP_MODE_SKB:
-		xdp_flags |= XDP_FLAGS_SKB_MODE;
-		break;
-	case XDP_MODE_NATIVE:
-		xdp_flags |= XDP_FLAGS_DRV_MODE;
-		break;
-	case XDP_MODE_HW:
-		xdp_flags |= XDP_FLAGS_HW_MODE;
-		break;
-	case XDP_MODE_UNSPEC:
-		break;
-	}
-
-	if (prog_name)
-		prog = bpf_object__find_program_by_title(obj, prog_name);
-	if (!prog)
-		prog = bpf_program__next(NULL, obj);
-
-	if (!prog) {
-		pr_warn("Couldn't find an eBPF program '%s' to attach. "
-			"This is a bug!\n", prog_name);
-		return -EFAULT;
-	}
-
-	prog_fd = bpf_program__fd(prog);
-	if (prog_fd <= 0) {
-		pr_warn("Invalid prog fd %d\n", prog_fd);
-		return -EFAULT;
-	}
-
-	err = bpf_set_link_xdp_fd(ifindex, prog_fd, xdp_flags);
-	if (err == -EEXIST && !(xdp_flags & XDP_FLAGS_UPDATE_IF_NOEXIST)) {
-		/* Force mode didn't work, probably because a program of the
-		 * opposite type is loaded. Let's unload that and try loading
-		 * again.
-		 */
-
-		__u32 old_flags = xdp_flags;
-
-		xdp_flags &= ~XDP_FLAGS_MODES;
-		xdp_flags |= (mode == XDP_MODE_SKB) ? XDP_FLAGS_DRV_MODE :
-			XDP_FLAGS_SKB_MODE;
-		err = bpf_set_link_xdp_fd(ifindex, -1, xdp_flags);
-		if (!err)
-			err = bpf_set_link_xdp_fd(ifindex, prog_fd, old_flags);
-	}
-	if (err < 0) {
-		pr_warn("Error attaching XDP program to %s: %s\n",
-			iface->ifname, strerror(-err));
-
-		switch (-err) {
-		case EBUSY:
-		case EEXIST:
-			pr_warn("XDP already loaded on device;"
-				" use --force to replace\n");
-			break;
-		case EOPNOTSUPP:
-			pr_warn("Native XDP not supported;"
-				" try using --skb-mode\n");
-			break;
-		default:
-			break;
-		}
-		return err;
-	}
-
-	pr_debug("Program '%s' loaded on interface '%s'%s\n",
-		 prog_name ?: bpf_program__title(prog, false),
-		 iface->ifname, mode == XDP_MODE_SKB ? " in skb mode" : "");
+	char pin_path[PATH_MAX];
+	int err = 0;
 
 	if (!pin_root_path)
-		return 0;
+		goto load;
 
 	err = make_dir_subdir(pin_root_path, "programs");
```

Comment:
Should we add an error message in this case? It might be useful if for some reason we can't create the pinned directory it gives an indication which directory and where we were trying to add it.